control/pid.go

```diff
@@ -126,40 +210,93 @@ func (p *basicPID) reset() error {
 	}
 
 	p.tuning = false
-	if p.kI == 0.0 && p.kD == 0.0 && p.kP == 0.0 {
-		var ssrVal float64
-		if p.cfg.Attribute["tune_ssr_value"] != nil {
-			ssrVal = p.cfg.Attribute["tune_ssr_value"].(float64)
-		}
+	if p.useMulti {
+		for i := 0; i < len(p.PIDSets); i++ {
 
-		tuneStepPct := 0.35
-		if p.cfg.Attribute.Has("tune_step_pct") {
-			tuneStepPct = p.cfg.Attribute["tune_step_pct"].(float64)
-		}
+			if p.PIDSets[i].I == 0.0 && p.PIDSets[i].D == 0.0 && p.PIDSets[i].P == 0.0 {
+				var ssrVal float64
+				if p.cfg.Attribute["tune_ssr_value"] != nil {
+					ssrVal = p.cfg.Attribute["tune_ssr_value"].(float64)
+				}
 
-		tuneMethod := tuneMethodZiegerNicholsPID
-		if p.cfg.Attribute.Has("tune_method") {
-			tuneMethod = tuneCalcMethod(p.cfg.Attribute["tune_method"].(string))
-		}
+				tuneStepPct := 0.35
+				if p.cfg.Attribute.Has("tune_step_pct") {
+					tuneStepPct = p.cfg.Attribute["tune_step_pct"].(float64)
+				}
 
-		p.tuner = pidTuner{
-			limUp:      p.limUp,
-			limLo:      p.limLo,
-			ssRValue:   ssrVal,
-			tuneMethod: tuneMethod,
-			stepPct:    tuneStepPct,
-		}
-		err := p.tuner.reset()
-		if err != nil {
-			return err
+				tuneMethod := tuneMethodZiegerNicholsPID
+				if p.cfg.Attribute.Has("tune_method") {
+					tuneMethod = tuneCalcMethod(p.cfg.Attribute["tune_method"].(string))
+				}
+
+				// Create a Tuner object for our PID set. Across all Tuner objects, they share global
+				// values (limUp, limLo, ssR, tuneMethod, stepPct). The only values that differ are P,I,D.
+				p.tuners[i] = &pidTuner{
+					limUp:      p.limUp,
+					limLo:      p.limLo,
+					ssRValue:   ssrVal,
+					tuneMethod: tuneMethod,
+					stepPct:    tuneStepPct,
+					kP:         p.PIDSets[i].P,
+					kI:         p.PIDSets[i].I,
+					kD:         p.PIDSets[i].D,
+				}
```

Comment:
would it be worth making a helper function for this code? since both the mimo and single input/output code use it. no is an okay answer since we eventually only want one code path

Reply:
leaving this code as is, as we will eventually remove the single input/output path once MIMO is ready